networking_aci/extensions/acioperations.py

```diff
@@ -0,0 +1,219 @@
+                hostgroups[hg_name] = ACI_CONFIG.get_hostgroup(hg_name)
+
+        return {
+            'hostgroups': hostgroups,
+            'fixed_bindings': ACI_CONFIG.fixed_bindings,
+            'address_scopes': ACI_CONFIG.address_scopes,
+        }
+        # return {"status": "success", "index": True, "arguments": kwargs}
+
+    def show(self, request, **kwargs):
+        """Show a single hostgroup"""
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        if request.GET.get("raw", False):
+            hostgroup = ACI_CONFIG.hostgroups.get(hostgroup_name)
+        else:
+            hostgroup = ACI_CONFIG.get_hostgroup(hostgroup_name)
+
+        if not hostgroup:
+            raise web_exc.HTTPNotFound('Hostgroup "{}" does not exist'.format(hostgroup_name))
+
+        return hostgroup
+
+
+class HostgroupModeController(wsgi.Controller):
+    def __init__(self, plugin, db, rpc_notifier):
+        super(HostgroupModeController, self).__init__()
+        self.plugin = plugin
+        self.db = db
+        self.rpc_notifier = rpc_notifier
+
+    def index(self, request, **kwargs):
+        check_cloud_admin(request)
+        ctx = context.get_admin_context()
+        return dict(hostgroup_modes=self.db.get_hostgroup_modes(ctx))
+
+    def show(self, request, **kwargs):
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        ctx = context.get_admin_context()
+        return dict(hostgroup_name=hostgroup_name, hostgroup_mode=self._get_mode(ctx, hostgroup_name))
+
+    def update(self, request, **kwargs):
+        check_cloud_admin(request)
+        hostgroup_name = kwargs.pop('id')
+        new_mode = kwargs.get("body", {}).get("mode")
+        if new_mode not in (aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL):
+            raise web_exc.HTTPBadRequest("Specify a mode that is either {} or {} (given mode was '{}')"
+                                         .format(aci_const.MODE_INFRA, aci_const.MODE_BAREMETAL, new_mode))
+
+        ctx = context.get_admin_context()
+        curr_mode = self._get_mode(ctx, hostgroup_name)
+        if curr_mode == new_mode:
+            raise web_exc.HTTPBadRequest("{} is already in {} mode".format(hostgroup_name, new_mode))
+
+        # check if there are still  on this hostgroup
+        hg_config = ACI_CONFIG.get_hostgroup(hostgroup_name)
+        if not hg_config:
+            raise web_exc.HTTPBadRequest("Hostgroup {} has no config associated with it".format(hostgroup_name))
+
+        for segment_id in self.db.get_segment_ids_by_physnet(ctx, hg_config['physical_network']):
+            hosts = self.db.get_hosts_on_segment(ctx, segment_id)
+            for host in hg_config['hosts']:
+                if host in hosts:
+                    LOG.error("Denied API request to switch hostgroup %s to %s mode, host %s still bound in segment %s",
+                              hostgroup_name, new_mode, host, segment_id)
```

Comment:
If you use set operations, you could return all the hosts being still problematic in a single reply, making the RTT for freeing up the hostgroup shorter. But just a suggestion and maybe not worth the effort/follow-up work.

Reply:
In nearly all cases we only will have one host in the hostgroup for direct_mode hostgroups, so I don't it's worth adding a "host(s)" to the commit message and making it therefore less beautiful. ;)